Simbody/src/CableSpan.cpp

```diff
@@ -0,0 +1,3028 @@
+        unitPower += ~unitForce_G * v;
+    }
+
+    {
+        calcUnitForceAtCableTermination(*this, state, unitForce_G);
+        SpatialVec v = getTerminationBody().getBodyVelocity(state);
+        unitPower += ~unitForce_G * v;
+    }
+
+    return unitPower * tension;
+}
+
+//==============================================================================
+//                             Path Error Vector
+//==============================================================================
+// This section contains helper functions for computing the path error vector.
+namespace
+{
+
+// Compute the straight line segments in between the obstacles.
+void calcLineSegments(
+    const CableSpan::Impl& cable,
+    const State& s,
+    const Vec3& pathOriginPoint,
+    const Vec3& pathTerminationPoint,
+    std::vector<LineSegment>& lines)
+{
+    lines.clear();
+    Vec3 prevPathPoint(pathOriginPoint);
+
+    cable.forEachActiveCurveSegment(
+        s,
+        [&](const CurveSegment& curve)
+        {
+            const CurveSegmentData::Pos& dataPos = curve.getDataPos(s);
+
+            // Compute the line segment from the previous point to the
+            // first curve contact point.
+            lines.emplace_back(prevPathPoint, dataPos.X_GP.p());
+
+            // The next line segment will start at the curve's final contact
+            // point.
+            prevPathPoint = dataPos.X_GQ.p();
+        });
+
+    // Compute the last line segment.
+    lines.emplace_back(prevPathPoint, pathTerminationPoint);
+}
+
+// Helper for computing a single element of the path error vector.
+// The algorithm for finding the correct path will attempt to drive this path
+// error to zero. Since we would like the curve's tangents to be aligned with
+// the straight line segments, the path error is defined as the alignment of the
+// line segments with the normal and binormal axes.
+// In that case: zero path error -> no alignment of line segments with normal
+// and binormal -> line segments must be aligned with the tangents.
+Real calcPathErrorElement(
+    const LineSegment& e,
+    const FrenetFrame& X,
+    CoordinateAxis axis)
+{
+    return dot(e.direction, X.R().getAxisUnitVec(axis));
+}
+
+// Compute the path error for each curve segment at the boundary frames, and
+// stack them in a vector. At each boundary frame (X_GP, X_GQ) the path error
+// along the given axes is computed, in that order.
+template <size_t N>
+void calcPathErrorVector(
+    const CableSpan::Impl& cable,
+    const State& s,
+    const std::vector<LineSegment>& lines,
+    std::array<CoordinateAxis, N> axes,
+    Vector& pathError)
+{
+    // Reset path error vector to zero.
+    pathError *= 0;
+
+    // The element in the path error vector to write to.
+    int pathErrorIx = -1;
+    // Iterator over the straight line segments connecting each curve segment.
+    auto lineIt = lines.begin();
+
+    cable.forEachActiveCurveSegment(
```

Comment:
Feels like you should use an index and range-check it, rather than use an unchecked iterator